### PR TITLE
SQL-1935: Implement do_auth_flow

### PR DESCRIPTION
**Pull request by pmeredit**
Note this can be tested by running `cargo run --bin do_auth_flow`, but you have to be on the vpn. I got lazy/impatient trying to figure out what was wrong with my Okta Auth0 config, so I'm just running this against the config from one of the OIDC enabled ADF instances. On the bright side, I believe it will work for anyone in the mongodb org due to this.

### CODE REVIEW

File: core/src/bin/do_auth_flow.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+use mongo_odbc_core::oidc_auth::*;
+
+#[tokio::main]
+async fn main() {
+    let c = CallbackContext {
```
Comment (pmeredit) [inline core/src/bin/do_auth_flow.rs:5]: Note, none of this stuff is sensitive.

File: rfc8252_http_server/src/lib.rs [hunk 2]
```diff
@@ -318,11 +318,13 @@ pub async fn tokio_start() -> (
 fn rfc8252_http_server_threaded_accepted() {
     use reqwest;
     let (server_handle, oidc_params_receiver) = threaded_start();
+
```
Comment (pmeredit) [inline rfc8252_http_server/src/lib.rs:321]:
```suggestion
```

File: rfc8252_http_server/src/lib.rs [hunk 3]
```diff
@@ -318,11 +318,13 @@ pub async fn tokio_start() -> (
...
     let _ = reqwest::blocking::get(format!(
         "{}{}",
         DEFAULT_REDIRECT_URI, "/callback?code=1234&state=foo"
     ))
     .unwrap();
+
```
Comment (pmeredit) [inline rfc8252_http_server/src/lib.rs:327]:
```suggestion
```

File: rfc8252_http_server/src/lib.rs [hunk 4]
```diff
@@ -385,7 +388,9 @@ async fn rfc8252_http_server_tokio_error() {
 #[tokio::test]
 async fn rfc8252_http_server_tokio_no_params() {
     let (server_handle, mut oidc_params_receiver) = tokio_start().await;
+
```
Comment (pmeredit) [inline rfc8252_http_server/src/lib.rs:391]:
```suggestion
```

File: rfc8252_http_server/src/lib.rs [hunk 5]
```diff
@@ -385,7 +388,9 @@ async fn rfc8252_http_server_tokio_error() {
...
     let _ = reqwest::get(DEFAULT_REDIRECT_URI).await.unwrap();
+
```
Comment (pmeredit) [inline rfc8252_http_server/src/lib.rs:393]:
```suggestion
```

File: core/src/oidc_auth.rs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+pub async fn do_auth_flow(params: CallbackContext) -> Result<IdpServerResponse, Error> {
+    let idp_info = params.idp_info.ok_or(Error::NoIdpServerInfo)?;
+    let client_id = idp_info.client_id;
+    let issuer_uri = IssuerUrl::new(idp_info.issuer).map_err(|e| Error::Other(e.to_string()))?;
+    if issuer_uri.url().scheme() != "https" {
+        return Err(Error::IssuerUriMustBeHttps);
+    }
+    let scopes = idp_info.request_scopes.unwrap_or_else(Vec::new);
+
```
Comment (pmeredit) [inline core/src/oidc_auth.rs:54]: I realized that we can actually use the tokio version here.

**Comment by pmeredit** [comment]
My commit statement that threaded will not work is actually untrue. It works fine because it's spawning an OS threaded within tokio. It is true, however, that it's being wasteful, we should let tokio spawn threads as necessary.

File: rfc8252_http_server/src/lib.rs [hunk 7]
```diff
@@ -190,7 +190,7 @@ async fn tokio_error(
     let _ = oidc_params_sender
         .send(Err(format!("{}: {}", error, error_description)))
         .await;
-    Ok(HttpResponse::build(http::StatusCode::OK)
```
Comment (pmeredit) [inline rfc8252_http_server/src/lib.rs:193]: this is a change from the 1934 PR, not sure how to get github to refresh

Comment (pmeredit) [reply rfc8252_http_server/src/lib.rs:193]: changing base didn't help it 🤷

**Comment by pmeredit** [comment]
As soon as SQL-1934 merges, I will rebase this to master.

File: core/src/oidc_auth.rs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    AuthorizationCode, ClientId, CsrfToken, IssuerUrl, Nonce, OAuth2TokenResponse,
+    PkceCodeChallenge, RedirectUrl, RequestTokenError, Scope,
+};
+use rfc8252_http_server::{start, OidcResponseParams};
+use std::{collections::HashSet, hash::RandomState, time::Instant};
+
+const DEFAULT_REDIRECT_URI: &str = "http://localhost:27097/redirect";
+
+// temporary until rust driver OIDC support is released
+// TODO Remove Me.
```
Comment (nbagnard) [inline core/src/oidc_auth.rs:14]: Do you have a JIRA to reference?

File: core/src/oidc_auth.rs [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+            };
+            Error::Other(format!("OpenID Connect: code exchange failed: {}", msg))
+        })?;
+
+    // Extract the auth and refresh tokens, and the expiration duration in seconds
+    let access_token = token_response.access_token().secret().to_string();
+    let refresh_token = token_response
+        .refresh_token()
+        .map(|t| t.secret().to_string());
+    let expires = token_response.expires_in();
```
Comment (nbagnard) [inline core/src/oidc_auth.rs:168]: [nit]
```suggestion
    let expires_in = token_response.expires_in();
```

**Review by nbagnard — APPROVED** [review]
LGTM!
The flow is super easy to follow, thank you for all the comments guiding us along :)

**Comment by pmeredit** [comment]
> LGTM. I did try testing the Okta login and also have trouble, I get `unable to sign in`.

You might need to be in the cloud-dev mana group

**Pull request merged**